src/main/java/com/fasterxml/jackson/core/filter/JsonPointerBasedFilter.java

```diff
@@ -13,25 +13,40 @@
 public class JsonPointerBasedFilter extends TokenFilter
 {
     protected final JsonPointer _pathToMatch;
+    protected final boolean _includeAllElements;
 
     public JsonPointerBasedFilter(String ptrExpr) {
-        this(JsonPointer.compile(ptrExpr));
+        this(JsonPointer.compile(ptrExpr), false);
     }
 
     public JsonPointerBasedFilter(JsonPointer match) {
+        this(match, false);
+    }
+
+    public JsonPointerBasedFilter(String ptrExpr, boolean includeAllElements) {
+        this(JsonPointer.compile(ptrExpr), includeAllElements);
+    }
+
+    public JsonPointerBasedFilter(JsonPointer match, boolean includeAllElements) {
         _pathToMatch = match;
+        _includeAllElements = includeAllElements;
     }
 
     @Override
     public TokenFilter includeElement(int index) {
-        JsonPointer next = _pathToMatch.matchElement(index);
+        JsonPointer next;
+        if (_includeAllElements && -1 == _pathToMatch.getMatchingIndex()) {
```

Comment:
Use `!_pathToMatch.mayMatchElement()` instead of checking `getMatchinIndex()` (same thing just bit more readable)